RFC: machines: Add LibvirtDBus provider

KKoukiou opened the pull request:
This patch series is rewriting https://github.com/martinpitt/cockpit/commit/329e62340ce9c2a6e8a7774b82677c8f294c503b and implements a second provider for Libvirt using Libvirt's D-Bus API (https://github.com/libvirt/libvirt-dbus/).
Since there is a lot of code that can be reused from libvirt.es6 file in the new provider, all reusable functions where moved to libvirt-common.es6 file to avoid code duplicity.

Antique commented on pkg/machines/index.es6:
```diff
@@ -51,3 +46,33 @@ export function appMain() {
+                client.close();
+                renderApp();
+            });
+}
```
I would change this code in a way that I would introduce a helper function called `checkLibvirtDBus()` or something like that and use that function in `appMain()` to decide what provider should be used.  That way there would be no need to create `renderApp()` function and have the call on multiple places.

KKoukiou commented:
@mareklibra @martinpitt @Antique  I reposted the patches with some fixes, adjusted the tests to run with both providers and some added info about libvirt-dbus in plugin docs.
Can you please take a look?

The tests are all passing locally, though with the current images they will fail here, because the available libvirt-dbus release is missing some APIs that I am using in these commits. (See more here https://www.redhat.com/archives/libvir-list/2018-June/msg00898.html)

So, although we shouldn't merge this patch series without the new release of libvirt-dbus to be included in the test images, can you please review so that I can get it to a state ready to merge?

Regarding supporting the using the libvirt-dbus in oVirt package, I suggest to keep it in separate PR, since current is already very big. WDYT?

KKoukiou commented:
Changes from previous push:
* Rebased on top on mlibra's PR for Adding disks, ht<span>tps://github.com/cockpit-project/cockpit/pull/9166
* Added support for VCPUS

martinpitt commented:
#9692 landed, can you please rebase?

KKoukiou commented on test/verify/check-machines-dbus:
```diff
@@ -0,0 +1,67 @@
+
+    def detectLibvirtDbus(self):
+        m = self.machine
+        namesList = m.execute("dbus-send --system --dest=org.freedesktop.DBus "
```
Not really, this code is running on the machine you execute test tests from, m.execute executes the commands inside the VM the tests are going to run inside. If I was using dbus python module I would need to run it from a script that was running inside machine that contains the dbus I want to poll.. Or correct me if I am wrong.

Antique replied:
Oh right, I missed the fact the it is executed on a different machine.

KKoukiou commented on test/verify/machineslib.py:
```diff
@@ -528,6 +524,8 @@ def testAddDisk(self):
         b.wait_present("#vm-subVmTest1-disks") # wait for the tab
         b.click("#vm-subVmTest1-disks") # open the "Disks" subtab
 
+        time.sleep(2);
```
I removed this already, I was  debugging something.

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,815 @@
+                return destroy().then(undefine);
+            } else {
+                return undefine();
+            }
```
Done

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,815 @@
+                                clientLibvirt[connectionName].call(storageVolPath[0], 'org.libvirt.StorageVol', 'Delete', [0], TIMEOUT)
+                                        .finally(undefine_synced.bind(null, i + 1));
+                            })
+                else /* VIR_DOMAIN_UNDEFINE_MANAGED_SAVE */
```
Done

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,815 @@
+        return dispatch => {
+            clientLibvirt[connectionName].call(objPath, 'org.libvirt.Domain', 'GetXMLDesc', [0], TIMEOUT)
+                    .done(domXml => {
+                        let updatedXml = updateNetworkIfaceState(domXml[0], networkMac, state);
```
Done

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,815 @@
+        connectionName
+    }) {
+        return dispatch => {
+            clientLibvirt[connectionName].call(objPath, 'org.libvirt.Domain', 'GetXMLDesc', [0], TIMEOUT)
```
Done

martinpitt replied:
Where is it checked? I don't see a `.catch` handler here?

KKoukiou commented on pkg/machines/index.es6:
```diff
@@ -24,24 +24,43 @@ import App from './app.jsx';
+                let providerVal = libvirtDBusavailable ? LibvirtDbus : Libvirt;
+
+                client.close();
+                return providerVal;
```
Done

KKoukiou commented on doc/guide/feature-virtualmachines.xml:
```diff
@@ -21,6 +25,15 @@
     </section>
 
+    <section id="feature-virtualmachines-sessionaccess">
+    <title>Managing VMs on session connection</title>
```
Done

martinpitt commented:
[testSerialConsole](https://fedorapeople.org/groups/cockpit/logs/pull-9240-20180827-193838-911deb24-verify-debian-stable/log.html#239) now fails across the board, on the well-known issue #9641. @KKoukiou , you need to update bots/naughty/*/9641-xtermjs-dimensions-crash for the changed path, as that test is now in machineslib.py instead of test/verify/check-machines.

martinpitt commented on test/verify/check-machines-dbus:
```diff
@@ -0,0 +1,67 @@
+                              "org.freedesktop.DBus.ListActivatableNames")
+
+        if ('org.libvirt' not in namesList):
+            raise unittest.SkipTest("libvirt-dbus is not available");
```
Might be a bit more obvious to do something like `busctl introspect org.libvirt /` (or whichever its main object path is), and just see if it fails. But this is okay as well, and just nitpicking.

KKoukiou replied:
I tired this, but I can't find a way to use cockpit dbus proxy call without specifying the interface but only the object path.
I would like to use this:
```proxy = client.proxy([interface, path], [options])```
but I should pass interface before path.

martinpitt replied:
Right, I actually just meant for the test here. As for `detectLibvirtProvider()`, if you want to change it there, don't use a proxy, that's too expensive. You just want to call a single method, like getting a property value or some `Ping()` like thing. I. e. use `client.call` there like you already do, just on some method on org.libvirt. In either case you have to specify an interface of the method or property, of course.

KKoukiou replied:
Oh I misread that it was the js code. Done.

martinpitt replied:
No blanket `except:` please. Use `except subprocess.CalledProcessError:` to only catch this specific one. We don't want to silently hide `AttributeError` or something similarly nasty/silly.

KKoukiou replied:
Done

martinpitt commented on test/verify/check-machines-dbus:
```diff
@@ -0,0 +1,67 @@
+            <allow send_destination="org.libvirt"/> \
+        </policy> \
+        </busconfig>'
+        m.execute("echo '{0}' > {1}".format(libvirtConfig, "/etc/dbus-1/system.d/org.libvirt.conf"))
```
This needs some explanation. admin can sudo and is a polkit admin too, so why doesn't libvirt-dbus'es polkit policy do the same as what `virsh -c qemu:///system` does? Or if it's really meant to be restricted to the root user (ugh, who does that), the tests should also reflect what happens in reality -- i. e. users wouldn't muck around with the policy file, but you would log in as a sudoer. And that's what the tests usually do. This looks like the code is missing a `{ superuser: "try" }` or even `"require"` somewhere?

KKoukiou replied:
So, in order to have access as a normal user to system libvirt dbus the user needs either this file or he needs to be part of the libvirt group.

The admin user in the images we are using is not part of the libvirt group, so I decided to do the change as a part of the test suite only.

Note that for the same reason (admin user is not a part of the libvirt group) he doesn't have access to the system libvirt with virsh either.

See output below from the debian-testing machine:
```
root@unassigned-hostname:~# id admin 
uid=1000(admin) gid=1000(admin) groups=1000(admin),27(sudo)
root@unassigned-hostname:~# su - admin
admin@unassigned-hostname:~$ virsh -c qemu:///system
==== AUTHENTICATING FOR org.libvirt.unix.manage ===
System policy prevents management of local virtualized systems
Authenticating as: Administrator (admin)
Password: 

```

KKoukiou replied:
So I can either leave it like this or change all images to add the admin user to libvirt group.

martinpitt replied:
No need to change the images -- just do the `usermod -a -G libvirt admin` call in the dbus test class'es `setUp()`. As long as you do this before `login_and_go()`, it will apply to the cockpit session.

KKoukiou replied:
Done. Just a note that this bahavior is available since libvirt-dbus v1.2.0.
debian-testing has this version, but current fedora-testing image  not.
fedora-* images should be updated with newer libvirt-dbus (and libvirt for the bugfix that got backported) before the tests get enabled for fedora.

martinpitt commented on test/verify/check-machines-dbus:
```diff
@@ -0,0 +1,67 @@
+
+        if ('org.libvirt' not in namesList):
+            raise unittest.SkipTest("libvirt-dbus is not available");
+        m.execute("echo 'PROVIDER=libvirt-dbus' > /etc/environment")
```
I think this is just a flag for machineslib.py, isn't it? It would IMHO be better to pass that as a class property in `class TestMachinesDBus`, and set `provider = "dbus"` here, and "virsh" in the other one, and then machineslib.py can just query that. This doesn't need to get all the way into the testbed.

KKoukiou replied:
TestMachinesDBus is the child class, and TestMachines is the parent class and we are inheriting the whole test suite from TestMachines.
Is it possible to query attributes of the child class from the parent class?

KKoukiou replied:
I just realized I can set the attribute provider to the parent class and just override it.
Done

martinpitt commented on test/verify/check-machines-dbus:
```diff
@@ -0,0 +1,67 @@
+        # enforce use of cockpit-machines instead of cockpit-machines-ovirt
+        m = self.machine
+        m.execute("sed -i 's/\"priority\".*$/\"priority\": 100,/' {0}".format("/usr/share/cockpit/machines/manifest.json"))
+        m.execute("[ ! -e {0} ] || sed -i 's/\"priority\".*$/\"priority\": 0,/' {0}".format("/usr/share/cockpit/ovirt/manifest.json"))
```
Doesn't that apply to the virsh test case as well? I. e. shuodln't that go/stay in machineslib.py?

KKoukiou replied:
Actually this part of the setup function is the same in the two classes, I 'll move it to the parent class.

KKoukiou replied:
Done

martinpitt commented on test/verify/check-machines-virsh:
```diff
@@ -0,0 +1,47 @@
+        # enforce use of cockpit-machines instead of cockpit-machines-ovirt
+        m = self.machine
+        m.execute("sed -i 's/\"priority\".*$/\"priority\": 100,/' {0}".format("/usr/share/cockpit/machines/manifest.json"))
+        m.execute("[ ! -e {0} ] || sed -i 's/\"priority\".*$/\"priority\": 0,/' {0}".format("/usr/share/cockpit/ovirt/manifest.json"))
```
Right, so it's doing the exact same thing here, please move it to `TestMachines.setUp()` and call the `super` constructor in the two test cases.

KKoukiou replied:
Done

martinpitt commented on bots/naughty/rhel-7/9641-xtermjs-dimensions-crash:
```diff
@@ -1,4 +1,4 @@
 Traceback (most recent call last):
-  File "test/verify/check-machines", line *, in testSerialConsole
+  File "test/verify/machineslib.py", line *, in testSerialConsole
```
Please squash that commit into the one that splits the test case. Also, it's "naughty", not "naughtly" :)

KKoukiou replied:
Done

martinpitt commented on test/verify/check-machines-dbus:
```diff
@@ -0,0 +1,50 @@
+    def allowSystemDBusUnprivileged(self):
+        m = self.machine
+        # Add admin user to libvirt group so that he has access to system libvirt-dbus
+        m.execute("sudo usermod -a -G libvirt admin")
```
This looks much better indeed. Also, defining `m` for just a single use is overkill :-) (Same for the helper function above).

KKoukiou replied:
Done

martinpitt commented on test/verify/check-machines-virsh:
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/python2
+# -*- coding: utf-8 -*-
```
This needs to be Python 3 now, and the `coding` is obsolete with py3.

KKoukiou replied:
Done

martinpitt commented on test/verify/check-machines-virsh:
```diff
@@ -0,0 +1,43 @@
+        if ("debian" in self.machine.image) or ("ubuntu" in self.machine.image):
+            m.execute("dpkg --remove libvirt-dbus || true")
+        else:
+            m.execute("rpm -e libvirt-dbus || true")
```
Do these commands regularly fail? If not, I suggest to remove the alternative truth, to fail more obviously if these ever get renamed, or our images get broken somehow. Otherwise this test case could silently test the dbus case again.

martinpitt replied:
Thanks, makes sense.

martinpitt commented on test/verify/machineslib.py:
```diff
@@ -341,7 +341,15 @@ def testBasic(self):
+        try:
+            with b.wait_timeout(1):
+                b.wait_visible("#vm-subVmTest2-run")
+        except:
```
No blanket exceptions. Also this is an extremely tight timeout. Thirdly, test cases  really shouldn't be written in this "opportunistic" and "branched" way. It should be predictable what happens at each step, so they should be linear. 

Either *enforce* that you get an exception (for cases that make sense), or in this case perhaps something like `b.wait_not_present("#vm-subVmTest2-run")` or perhaps `wait_not_visible()` makes more sense?

If there's a good reason for it (I don't see it, I'm not that familiar with what happens here), please add a comment.

KKoukiou replied:
I was trying to debug the testBasic failure. It should not be in the final code.

martinpitt commented:
I added a TODO item to the checklist in the description about adjusting the dependencies (in tools/cockpit.spec and tools/debian/control). This should now prefer libvirt-dbus, and fall back to libvirt-client, IMHO.

martinpitt commented on pkg/machines/actions/provider-actions.es6:
```diff
@@ -59,8 +59,8 @@ import {
-export function attachDisk({ connectionName, diskFileName, target, permanent, hotplug, vmName }) {
-    return virt(ATTACH_DISK, { connectionName, diskFileName, target, permanent, hotplug, vmName });
+export function attachDisk({ connectionName, diskFileName, target, permanent, hotplug, vmName, vmId }) {
+    return virt(ATTACH_DISK, { connectionName, diskFileName, target, permanent, hotplug, vmName, vmId });
```
It might make sense to split out the "vmID" changes to a separate PR that can already land. That's the kind of preparatory PR that is rather invasive (thus bitrots fast) but also doesn't/shouldn't change functionality much, so you can land it fast.

KKoukiou replied:
Done

martinpitt commented on pkg/machines/libvirt.es6:
```diff
@@ -150,7 +150,7 @@ LIBVIRT_PROVIDER = {
      * @returns {Function}
      */
-    GET_VM ({ lookupId: name, connectionName }) {
+    GET_VM ({ name, connectionName }) {
```
Perhaps this file should now be renamed to libvirt-virsh.es6, for symmetry with libvirt-dbus.es6 and to avoid confusion?

Also, everything here is still name based, whereas -dbus uses Id, and the provider-actions get both. This is a bit confusing as well, and the provider action could potentially get inconsistent name and ID. It  might be clearer to just declare the name being the "id" here, and rename everything "id"? It should be treated as an opaque, backend-defined value, so the UI should not care about it and just pass it around, and the actions would then only have one ID.

KKoukiou replied:
libvirt.es6 renaming to libvirt-virsh.es6 done.

Regarding the id comment I have some worries.

* In case we are calling the actions from the react component, we should not deal with what is the underlying implementation. 
See here:
https://github.com/cockpit-project/cockpit/blob/4edbf97298bc85fb4aa17196b52cbd110dc1e79c/pkg/machines/components/diskAdd.jsx#L424
If we follow the pattern you suggest, and call it like getVm({id: vm.name, connectionName}) this will work only for the libvirt-virsh and with id it will work only for dbus provider. It's of course possible to find out what provider is used, but it's not correct to check it there I think.
 * In libvirt-dbus provider methods, we need mostly both name and id attributes in the parameters. Id to use it the API calls and name to print errors to the user if something went bad. User probably wants to see name in the error message and not VM id.

So, considering these I think it makes sense to keep both attributes passed to each action, even though sometimes they are not used. Also, I am trying to use everywhere named args, so that order of parameters will not matter.

martinpitt replied:
I see, thanks for the explanation! Makes sense. (Good to discuss this stuff and have a written record of it :smile: )

martinpitt commented on pkg/machines/reducers.es6:
```diff
@@ -160,13 +160,22 @@ function vms(state, action) {
+        else
+            return state
+                    .filter(vm => (action.connectionName !== vm.connectionName || action.name != vm.name ||
+                        (action.transientOnly && vm.persistent)));
```
.. the above (unifying name and id) should also avoid these nasty case handling, i. e. the backend specifics leaking into the abstract reducers.

KKoukiou replied:
Resolving as it's the same as what we discussed above.

martinpitt commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,861 @@
+ * See https://github.com/libvirt/libvirt-dbus
+ */
+import cockpit from 'cockpit';
+import $ from 'jquery';
```
Please avoid introducing jQuery into new code. It's only being used once for the `isEmptyObject()` function. I'll write there how to replace it.

martinpitt replied:
The thing to avoid is to mix React and jQuery UI building (e. g. show/hide elements), as they are both colliding technically and also makes it hard to understand as a developer what a page is doing.

For XML parsing, doesn't [DOMParser](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser) work? It would avoid pulling the sizable jQuery into the page just for that one function. Otherwise, if DOMParser does not work, using jQuery for XML parsing would be acceptable.

KKoukiou replied:
DOMParser works, I replaced all occurrences in machines code.

martinpitt commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,861 @@
+                                .done((storageVolumePath) => {
+                                    clientLibvirt[connectionName].call(storageVolumePath[0], "org.freedesktop.DBus.Properties", "GetAll", ["org.libvirt.StorageVol"], TIMEOUT)
+                                            .done(function(props) {
+                                                if (!$.isEmptyObject(props[0]))
```
Please replace with non-jQuery alternative: `Object.getOwnPropertyNames(props[0]).length > 0`

KKoukiou replied:
Done

Antique replied:
So in this case we only need volume path, I would suggest using Get method instead of GetAll.  In addition wherever we use GetAll this check is not sufficient enough because the returned props can contain only subset of all available properties so it may happen that the one we need is not included and the condition would still be true.  GetAll is best effort dbus call and it doesn't have to return all properties, it returns what it can.

KKoukiou replied:
Done

martinpitt commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,861 @@
+                    .done(domXml => {
+                        let updatedXml = updateNetworkIfaceState(domXml[0], networkMac, state);
+                        if (!updatedXml) {
+                            dispatch(vmActionFailed({
```
.. so I suppose that's one half of the failure handler, if the dbus call succeeds, but returns an empty XML. But it could also fail, no?

KKoukiou replied:
Fixed

martinpitt commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,861 @@
+                    clientLibvirt[connectionName].call('/org/libvirt/QEMU', 'org.libvirt.Connect', 'StorageVolLookupByPath', [options.storage[i]], TIMEOUT)
+                            .done((storageVolPath) => {
+                                clientLibvirt[connectionName].call(storageVolPath[0], 'org.libvirt.StorageVol', 'Delete', [0], TIMEOUT)
+                                        .finally(undefine_synced.bind(null, i + 1));
```
This looks a bit curious. Deleting the storage volumes in parallel isn't safe? Or do they depend on each other somehow? If so, is it guaranteed that the dependencies are in numerical order? This could use a comment.

KKoukiou replied:
We can delete storage volumes in parallel. I just need to make sure that after deleting the storage all storage volumes I 'll Undefine the VM. You suggest putting all promises for volume deletion in some array and calling  Promise.al? Or what is the proper way to do it?

martinpitt replied:
Right, that's the usual approach. It also (usually) makes error handling easier, as `Promise.all()` has a sensible behaviour (succeeds if everything succeeds, or fails if anything fails)

KKoukiou replied:
ok thanks and sorry for my ignorance, javascript total newbie here.

martinpitt replied:
FTR, this can be written more elegantly using `promises = mylist.map()`, but this isn't a blocker.

martinpitt commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,861 @@
+    /* Check stopped libvirtd */
+    let systemdClient = cockpit.dbus('org.freedesktop.systemd1', { bus: connectionName });
+    systemdClient.subscribe(
+        { interface: 'org.freedesktop.systemd1.Manager', member: 'JobRemoved' },
```
I'm ok with this, but rather than introducing assumptions and knowledge about systemd here, it seems to me it would be more elegant and straightforward to monitor when the `clientLibvirt[connectionName]` object goes away. You can use the [track option of cockpit.dbus()](https://cockpit-project.org/guide/latest/cockpit-dbus.html) for that and subscribe to the "close" signal.

This will then work in the same way for the user libvirt instance, which is (most likely) not started by systemd, but on demand by the library or the session bus.

KKoukiou replied:
I tried the following but the hander gets never triggered:
```
-    let systemdClient = cockpit.dbus('org.freedesktop.systemd1', { bus: connectionName });
-    systemdClient.subscribe(
-        { interface: 'org.freedesktop.systemd1.Manager', member: 'JobRemoved' },
-        (path, iface, signal, args) => {
-            logDebug(`signal on ${path}: ${iface}.${signal}(${JSON.stringify(args)})`);
-
-            if (args[2] == 'libvirtd.service') {
-                dispatch(checkLibvirtStatus(libvirtServiceName));
-                dispatch(deleteUnlistedVMs(connectionName, []));
-                dispatch(delayPolling(getAllVms(connectionName, libvirtServiceName)));
-            }
-        });
+    cockpit.dbus('org.libvirt', { bus: connectionName, "track": true }).addEventListener(
+        "close",
+        () => {
+            logDebug("Libvirt went away from D-Bus");
+
+            dispatch(checkLibvirtStatus(libvirtServiceName));
+            dispatch(deleteUnlistedVMs(connectionName, []));
+            dispatch(delayPolling(getAllVms(connectionName, libvirtServiceName)));
+        }
+    );
```

Did you mean something different?
I tried to trigger it by restarting libvirtd or kill the libvirt-dbus process but this handler doesn't execute.

martinpitt replied:
Did you literally use this? That's not the cockpit.dbus() client you are really using, but an entirely new instance, isn't it? I suppose that just immediately gets GCed away again. Can you try on `clientLibvirt[connectionName]`?

We do that exact thing in the [packagekit lib](https://github.com/cockpit-project/cockpit/blob/master/pkg/lib/packagekit.es6#L70), but on a long-lived global variable, where it works just fine (we have a test for that).

KKoukiou replied:
So indeed it works, I just misunderstood what the signal is doing. So, it the existing code I am getting the signal when libvirtd is not running, thus I am able to run the checkLibvirtStatus and display a page showing 'Libvirtd service is not active'...
When I am listening to the libvirt-dbus close signal, I just get triggered when libvirt-dbus process existed which is not attached to libvirtd. libvirtd can be stopped, but the libvirt-dbus process running, so I will not get notified for a killed libvirtd. Does this make any sense?
Do you have any other idea how to track stopped libvirtd apart from the current implementation?
Probably I should also keep the EventListener to close signal on dbus libvirt. since they serve slightly different purpose. 

Note: I can't change the behavior of tracking libvirtd status changes because I bind to have same behavior are libvirt-virsh.js implementation.

martinpitt replied:
Oh right, I didn't consider that libvirt.service != libvirt-dbus. So I suppose this is okay for now. But how do you track the user instance then?

KKoukiou replied:
I don't track it actually. I couldn't think of some way to track it. Any suggestions?

KKoukiou replied:
BTW I added the EventListener to close signal for libvirt-dbus as well. I think it makes sence to have both libvirtd and libvirt-dbus close handling.

Antique replied:
I'm not sure whether it make sense to track libvirt-dbus.  Since it's a dbus daemon it gets started on demand and if it crashes it will be started again once we do some dbus call.  The only way how it can disappear is that it's removed from system completely.

We don't need to track the session (user) instance as well, the libvirtd for session instance is again started on demand, if it crashes/is killed it will be started again unless it's removed from system completely.

martinpitt replied:
OK, good to know. I was about to suggest that libvirt-dbus should notice when libvirt goes down, and die with it -- then libvirt can transitively be tracked through the D-Bus "close" signal. But it seems that this isn't actually necessary. Let's not dwell on this then -- if there is some fine-tuning to be made, it can become a follow-up.

KKoukiou replied:
Just for the review ease, I removed EventListener to close signal according to this discussion.

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,861 @@
+                                                                dispatch(getVm({connectionName, id:vmId}));
+                                                            });
+                                            })
+                                            .fail(function(ex) { console.warn("failed waiting fot the StorageVol proxy to get ready", ex) });
```
All these console warns and failures should be replaced with exceptions visible to UI -> dispatch(vmActionFailed...

KKoukiou replied:
Done.
However, now UI error appear in the Vm Overview subtab. Ideally it should be viewed on the subtab where they where actually caused.
NOTE: libvirt-virsh implementation doesn't show any exceptions to user from attach-disk and create-and-attach-disk

martinpitt commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,861 @@
+
+                    logDebug(`GET_STORAGE_POOLS: object paths: ${JSON.stringify(objPaths)}`);
+
+                    function get_storage_pool(i) {
```
The clean way to avoid that in the code is to wait with the `.setState()` call until all parallel promises are finished (using `Promise.all`). If that is somehow unwieldy, the test should have some appropriate `wait_present()` statements (e. g. wait for the n-th row in the table - the test should be able to predict how many rows it expects)

KKoukiou replied:
So, testing can be trickier sometimes;
I have a test that it's sometimes failing and basically does the following;

* click Add Disk to existing VM
* Click Use Existing Disk
* Choose Storage Pool
* Click the dropdown button to choose storage Volume.
Errors on storage volumes list being empty for specific pool.

I can never reproduce manually.
The issue that probably I hit here is that the test is getting to the button to click the dropdown for volumes before the state is updated so that it contains the storage volumes list.
The dropdown button will never be updated by react while it's still clicked.
So ```wait_present()``` will not help.

martinpitt commented:
You asked on IRC, but then went away:  The correct dependency would be `(libvirt-dbus or libvirt-client)`, then you shold be able to uninstall libvirt-dbus to test the virsh backend.

KKoukiou commented:
@martinpitt So the cockpit-machine package on debian depends on libvirt-daemon-system which internally depends on libvirt-clients. I think putting one more line there (libvirt-dbus or lbivirt-clients) will results in libvirt-dbus never being installed since libvirt-cilents dependency is brought by libvirt-daemon-system.
Also libvirt-dbus implementation still need lbivirt-clients because some operations still use virsh. aka creating VM. @mareklibra had implemented a shell script for VM creation/installation which he was called from javascript, and I didn't discard it's usage yet in libvirt-dbus implementation. So I basically need libvirt-client package in cockpit-machines libvirt-dbus implmentation as well.

martinpitt commented:
Removing needswork for now. I'll re-review that and possibly do some minor tweaks.

KKoukiou commented:
Failing tests which are not related to this PR:
testAddDisk: I opened issue about it https://github.com/cockpit-project/cockpit/issues/9945
testSerialConsole: Known issue
testCreate: Inconsistently fails with `CPU mode 'custom' for x86_64 kvm domain on x86_64 host is not supported by hypervisor`. I suspect it's related to the server where the tests are running.
from IRC: 
```
(03:57:45 PM) pitti: kkoukiou: well, I think some are on RHEL 7.4 still, some on 7.5
(04:08:46 PM) phrdina: pitti: so the only difference is rhel version and that could be the root 
cause, the issue is that libvirt will not get list of cpu models supported by qemu
```

KKoukiou commented on pkg/machines/index.es6:
```diff
@@ -51,3 +69,16 @@ export function appMain() {
+/**
+ * Start the application.
+ */
+export function appMain() {
```
Done

KKoukiou replied:
you meant detectLibvirtProvider, I fixed it, thanks

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,905 @@
+/* Default timeout for libvirt-dbus method calls */
+const TIMEOUT = { timeout: 30000 };
+
+export const Enum = {
```
Done

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,905 @@
+                        call(connectionName, storagePoolPath[0], 'org.libvirt.StoragePool', 'StorageVolCreateXML', [volXmlDesc, 0], TIMEOUT)
+                                .done((storageVolumePath) => {
+                                    call(connectionName, storageVolumePath[0], "org.freedesktop.DBus.Properties", "Get", ["org.libvirt.StorageVol", "Path"], TIMEOUT)
+                                            .done(function(props) {
```
Done

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,905 @@
+                    );
+                }
+                return Promise.all(storageVolPathsPromises);
+            })
```
Indentation is ok here.

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,905 @@
+        if (connectionName) {
+            return dispatch => {
+                dispatch(checkLibvirtStatus(libvirtServiceName));
+                clientLibvirt[connectionName] = dbus_client(connectionName);
```
Correct.

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,905 @@
+            }
+        });
+
+    /* Listen on a stopped libvirtd on systemd D-Bus. If user is using libvirtd not started
```
Done

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,905 @@
+    systemdClient.subscribe(
+        { interface: 'org.freedesktop.DBus.Properties', path: '/org/freedesktop/systemd1/unit/libvirtd_2eservice', member: 'PropertiesChanged' },
+        (path, iface, signal, args) => {
+            if (args[0] == "org.freedesktop.systemd1.Unit" && args[1].ActiveState.v == "deactivating") {
```
Done

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,905 @@
+/**
+ * Call a Libvirt method
+ */
+export function call(connectionName, objectPath, iface, method, args, opts) {
```
Done

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,905 @@
+                    logDebug(`GET_STORAGE_POOLS: object paths: ${JSON.stringify(objPaths)}`);
+
+                    for (let i = 0; i < objPaths[0].length; i++) {
+                        storagePoolsPropsPromises.push(call(connectionName, objPaths[0][i], "org.freedesktop.DBus.Properties", "GetAll", ["org.libvirt.StoragePool"], TIMEOUT));
```
Done

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,905 @@
+                                                 */
+                                                props.name = ("Name" in returnProps[0]) ? returnProps[0].Name.v.v : undefined;
+                                                props.persistent = ("Persistent" in returnProps[0]) ? returnProps[0].Persistent.v.v : undefined;
+                                                props.autostart = ("Autostart" in returnProps[0]) ? returnProps[0].Autostart.v.v : undefined;
```
ok

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,815 @@
+    if (interfaceElems) {
+        for (let i = 0; i < interfaceElems.length; i++) {
+            const interfaceElem = interfaceElems[i];
+            const macElem = getSingleOptionalElem(interfaceElem, 'mac');
```
fixed it, sorry for skipping it.

KKoukiou commented on pkg/machines/libvirt-dbus.es6:
```diff
@@ -0,0 +1,815 @@
+            if (mac !== networkMac)
+                continue;
+
+            let linkElem = getSingleOptionalElem(interfaceElem, 'link');
```
fixed. sorry for skipping again :(

martinpitt commented:
Blocking on #9952 to fix adding disks.

martinpitt commented:
I rebased this on top of master to fix the conflict, and on top of #9952 to get the updated test case. I also did some squashing and typographic fixes in the commit logs.

martinpitt commented:
The commit to skip tests on fedora-28 was garbled, I fixed that locally. Also, PR #9972 updates to libvirt 4.1.0-5.fc28 , so once that's in the skip can be dropped completely.

martinpitt commented:
In my local case, I get this failure:
```
  File "/home/martin/upstream/cockpit/test/verify/machineslib.py", line 627, in testAddDisk
    b.wait_not_present("#vm-subVmTest1-disks-vde-target")
```
I. e. after force-shutting down the VMs, it is supposed to retain the two *_permanent volumes, and lose the two *_temporary ones. But vde (mydiskofpoolone_temporary) is still present.

On CI, it fails a bit earlier, the "vdd" disk is not actually appearing.

All these failures are *not* a bug in the tests - I can reproduce this behaviour manually perfectly well, and after shutting down the temporary ones are still present in `virsh dumpxml subVmTest1`.

martinpitt commented:
I initially thought this was becase `AddDisk.onValueChanged("permanent", value)` was never being called when not changing the checkbox. Indeed it isn't (that seems like a bug, as the libvirsh default seems to be "permanent"). But it's still permanent even when I tick and untick the "attach permanently" box (then `onValueChanged` is called correctly).

Can it actually be seen in `dumpXml` whether a disk is permanent  or temporary? They look exactly the same here.

martinpitt commented:
When I ticked and unticked "Attach permanently", I also got this error:

ATTACH_DISK failed for diskFileName undefined {"problem":null,"name":"org.libvirt.Error","message":"Cannot access storage file 'undefined': No such file or directory"}

This isn't reliably reproducible. Somewhere in the XML building/property updates there's a major bug still :(

martinpitt commented:
Updated fedora-28 image landed with the libvirt fix, so rebased and dropped the skip.

KKoukiou commented:
@martinpitt  regarding the AddDisk failure indeed looks like I messed up with the libvirt flags for permanent/temporary change:
applying this fix:
```
-    VIR_DOMAIN_AFFECT_CONFIG: 0,
-    VIR_DOMAIN_AFFECT_CURRENT: 1,
-    VIR_DOMAIN_AFFECT_LIVE: 2,
+    VIR_DOMAIN_AFFECT_CURRENT: 0,
+    VIR_DOMAIN_AFFECT_LIVE: 1,
+    VIR_DOMAIN_AFFECT_CONFIG: 2,
```

martinpitt commented:
@KKoukiou : Gee, well spotted, thanks! Removing needswork label for now, let's see if things are happy now.

martinpitt commented:
- [fedora-28](https://209.132.184.41:8493/logs/pull-9240-20180917-115858-87ad9d9c-verify-fedora-28/log.html#248) failure: You need to update the 9641-xtermjs-dimensions-crash naughty override for all other OSes as well.
 - [rhel-x](https://209.132.184.41:8493/logs/pull-9240-20180917-121313-87ad9d9c-verify-rhel-x/log.html) image prep failure: libvirt-dbus does not exist on current rhel-x image. There were PRs to add it to Fedora and debian-testing, but not rhel-x. In order to not block on this, could the dependency only be applied on Fedora for now?

KKoukiou commented:
>     * [fedora-28](https://209.132.184.41:8493/logs/pull-9240-20180917-115858-87ad9d9c-verify-fedora-28/log.html#248) failure: You need to update the 9641-xtermjs-dimensions-crash naughty override for all other OSes as well.
> 
>     * [rhel-x](https://209.132.184.41:8493/logs/pull-9240-20180917-121313-87ad9d9c-verify-rhel-x/log.html) image prep failure: libvirt-dbus does not exist on current rhel-x image. There were PRs to add it to Fedora and debian-testing, but not rhel-x. In order to not block on this, could the dependency only be applied on Fedora for now?

Done

martinpitt commented:
@KKoukiou : I just landed #9945, and now this conflicts. Can you please rebase?

KKoukiou commented:
> @KKoukiou : I just landed #9945, and now this conflicts. Can you please rebase?

Done

KKoukiou commented:
> Related test failures:
> 
>     * testAddDisk on [debian-testing](https://209.132.184.41:8493/logs/pull-9240-20180917-191010-eb76890f-verify-debian-testing/log.html#104) and [fedora-29](https://209.132.184.41:8493/logs/pull-9240-20180917-191212-eb76890f-verify-fedora-29/log.html#104)

Looks like race condition when tests are executed very fast with state not being up to date.. I added some logs in react code and I can't reproduce locally any more.

> 
>     * testLibvirt on [fedora-29](https://209.132.184.41:8493/logs/pull-9240-20180917-191212-eb76890f-verify-fedora-29/log.html#111)

Not reproducible locally. I tried 10 times with no error :/

> 
>     * testCreate on [fedora-i386](https://209.132.184.41:8493/logs/pull-9240-20180917-190909-eb76890f-verify-fedora-i386/log.html#49)

This test can't be affected by this PR because the code for CREATE action is shared 1000% between the two providers.
I tried to clean a bit however the test steps and let's see it it can reappear.
> 
> 
> There are also some failures in the virsh tests, but these are (presumably) not new. But we shouldn't introduce more flakes, it's hard enough to get the tests to green as it is..

martinpitt commented:
Hmkay, retrying the tests a few times.

By the way, if you want to add libvirt-dbus to the rhel-x image, now would be a good time for it. I just  spent my entire morning yesterday to fix the current regressions, so we got a fresh build yesterday. There's a very high chance that a build today will also succeed :-)

martinpitt commented:
I can locally reproduce [this failure on fedora-29](https://209.132.184.41:8493/logs/pull-9240-20180918-102020-63a3c7cb-verify-fedora-29/log.html#225) in about one out of three (parallel) runs. Just keeping the log pointer here for now.

KKoukiou commented:
Reposting with the following diff. (forgot to change some naughy files)
```
[kkoukiou@localhost cockpit]$ git diff
diff --git a/bots/naughty/rhel-7/7891-selinux-denies-virtlogd-dbus-2 b/bots/naughty/rhel-7/7891-selinux-denies-virtlogd-dbus-2
index cf5808fb7..292a9aa0f 100644
--- a/bots/naughty/rhel-7/7891-selinux-denies-virtlogd-dbus-2
+++ b/bots/naughty/rhel-7/7891-selinux-denies-virtlogd-dbus-2
@@ -1,5 +1,5 @@
 Traceback (most recent call last):
-  File "test/verify/check-machines", line *, in testBasic
+  File "test/verify/machineslib.py", line *, in testBasic
     wait(lambda: "Linux version" in self.machine.execute("cat {0}".format(args["logfile"])), delay=3)
 *
 Error: Condition did not become true.
diff --git a/bots/naughty/rhel-x/7891-selinux-denies-virtlogd-dbus-2 b/bots/naughty/rhel-x/7891-selinux-denies-virtlogd-dbus-2
index cf5808fb7..292a9aa0f 100644
--- a/bots/naughty/rhel-x/7891-selinux-denies-virtlogd-dbus-2
+++ b/bots/naughty/rhel-x/7891-selinux-denies-virtlogd-dbus-2
@@ -1,5 +1,5 @@
 Traceback (most recent call last):
-  File "test/verify/check-machines", line *, in testBasic
+  File "test/verify/machineslib.py", line *, in testBasic
     wait(lambda: "Linux version" in self.machine.execute("cat {0}".format(args["logfile"])), delay=3)
 *
 Error: Condition did not become true.
```

martinpitt commented:
I now added `set +` to `install_machine.sh`, which is the one that is supposed to fail. And it does:
```
warning: GetXMLDesc method failed for path /org/libvirt/QEMU/domain/_2998e19d_220f_4d9c_8be8_b473b0dbdd58 
warning: + VM_NAME=subVmTestCreate15
+ SOURCE=/var/lib/libvirt/novell.iso
+ OS=macosx10.5
+ MEMORY_SIZE=921600
+ VCPUS=1
+ DISKS='path=/home/admin/.local/share/libvirt/images/subVmTestCreate15.qcow2,device=disk,bus=ide,readonly=off,shareable=off,driver_name=qemu,driver_type=qcow2
'
+ DISPLAYS='type=spice,listen=127.0.0.1
type=vnc,listen=127.0.0.1,port=-1

'
+ '[' -z macosx10.5 ']'
+ '[' -z 'path=/home/admin/.local/share/libvirt/images/subVmTestCreate15.qcow2,device=disk,bus=ide,readonly=off,shareable=off,driver_name=qemu,driver_type=qcow2
' ']'
+ createOptions 'path=/home/admin/.local/share/libvirt/images/subVmTestCreate15.qcow2,device=disk,bus=ide,readonly=off,shareable=off,driver_name=qemu,driver_type=qcow2
' --disk
+ CREATE_OPTIONS_RESULT=
+ IFS=
+ read -r PARAM
+ '[' -n '    path=/home/admin/.local/share/libvirt/images/subVmTestCreate15.qcow2,device=disk,bus=ide,readonly=off,shareable=off,driver_name=qemu,driver_type=qcow2' ']'
+ CREATE_OPTIONS_RESULT=' --disk     path=/home/admin/.local/share/libvirt/images/subVmTestCreate15.qcow2,device=disk,bus=ide,readonly=off,shareable=off,driver_name=qemu,driver_type=qcow2'
+ IFS=
+ read -r PARAM
+ '[' -n '' ']'
+ IFS=
+ read -r PARAM
+ DISKS_PARAM=' --disk     path=/home/admin/.local/share/libvirt/images/subVmTestCreate15.qcow2,device=disk,bus=ide,readonly=off,shareable=off,driver_name=qemu,driver_type=qcow2'
+ '[' -z 'type=spice,listen=127.0.0.1
type=vnc,listen=127.0.0.1,port=-1

' ']'
+ createOptions 'type=spice,listen=127.0.0.1
type=vnc,listen=127.0.0.1,port=-1

' --graphics
+ CREATE_OPTIONS_RESULT=
+ IFS=
+ read -r PARAM
+ '[' -n '    type=spice,listen=127.0.0.1' ']'
+ CREATE_OPTIONS_RESULT=' --graphics     type=spice,listen=127.0.0.1'
+ IFS=
+ read -r PARAM
+ '[' -n type=vnc,listen=127.0.0.1,port=-1 ']'
+ CREATE_OPTIONS_RESULT=' --graphics     type=spice,listen=127.0.0.1 --graphics type=vnc,listen=127.0.0.1,port=-1'
+ IFS=
+ read -r PARAM
+ '[' -n '' ']'
+ IFS=
+ read -r PARAM
+ '[' -n '' ']'
+ IFS=
+ read -r PARAM
+ GRAPHICS_PARAM=' --graphics     type=spice,listen=127.0.0.1 --graphics type=vnc,listen=127.0.0.1,port=-1'
++ cut -c 1
++ echo /var/lib/libvirt/novell.iso
+ FIRST_1_SOURCE=/
++ cut -c -3
++ echo /var/lib/libvirt/novell.iso
+ FIRST_3_SOURCE=/va
++ cut -c -4
++ echo /var/lib/libvirt/novell.iso
+ FIRST_4_SOURCE=/var
+ '[' / = / -a '!' -f /var/lib/libvirt/novell.iso ']'
+ '[' / = / -o /var = http -o /va = ftp -o /va = nfs ']'
+ LOCATION_PARAM='--cdrom /var/lib/libvirt/novell.iso'
++ mktemp
+ DOMAIN_FILE=/tmp/tmp.z6auzHq3sj
+ virsh -q destroy subVmTestCreate15
+ virsh -q dumpxml subVmTestCreate15
+ virsh -q undefine subVmTestCreate15 --managed-save
+ virt-install --name subVmTestCreate15 --os-variant macosx10.5 --memory 921600 --vcpus 1 --quiet --wait -1 --noautoconsole --noreboot --disk path=/home/admin/.local/share/libvirt/images/subVmTestCreate15.qcow2,device=disk,bus=ide,readonly=off,shareable=off,driver_name=qemu,driver_type=qcow2 --cdrom /var/lib/libvirt/novell.iso --graphics type=spice,listen=127.0.0.1 --graphics type=vnc,listen=127.0.0.1,port=-1
ERROR    internal error: qemu unexpectedly closed the monitor: 2018-09-18T13:26:46.201467Z qemu-system-x86_64: cannot set up guest memory 'pc.ram': Cannot allocate memory
Domain installation does not appear to have been successful.
If it was, you can restart your domain by running:
  virsh --connect qemu:///session start subVmTestCreate15
otherwise, please restart your installation.
+ EXIT_STATUS=1
+ '[' 1 -eq 0 ']'
+ vmExists subVmTestCreate15
+ grep -q --line-regexp --fixed-strings subVmTestCreate15
+ virsh list --all
+ awk '{print $2}'
+ virsh -q define /tmp/tmp.z6auzHq3sj
+ rm -f /tmp/tmp.z6auzHq3sj
+ exit 1
```
I noticed someting interesting in the journal:
```
Sep 18 09:28:07 localhost.localdomain systemd-coredump[3728]: Process 1640 (libvirtd) of user 1000 dumped core.
                                                              
                                                              Stack trace of thread 3726:
                                                              #0  0x00007fcbd0ee5da4 __pthread_mutex_lock (libpthread.so.0)
                                                              #1  0x00007fcbc8f924d2 virQEMUDriverGetConfig (libvirt_driver_qemu.so)
                                                              #2  0x00007fcbc8fd6ea0 n/a (libvirt_driver_qemu.so)
                                                              #3  0x00007fcbd188f99f virStateStop (libvirt.so.0)
```
That might or might not be related. Are these tests covering system or user qemu?

martinpitt commented:
Not [fully perfect](https://209.132.184.41:8493/logs/pull-9240-20180918-163131-378e000e-verify-debian-testing/log.html#106) yet, but this looks *so* much better, thanks!
